pkg/traceability/muleemitter.go

```diff
@@ -63,18 +73,49 @@ func (me *MuleEventEmitter) Start() error {
 		return err
 	}
 
+	var lastTime time.Time
+	lastTime, err = time.Parse(time.RFC3339, strStartTime)
+	if err != nil {
+		logrus.WithError(err).Error("Unable to Parse Last Time")
+		return err
+	}
 	for _, event := range events {
+		// Results are not sorted. We want the most recent time to bubble up
+		if event.Timestamp.After(lastTime){
+			lastTime = event.Timestamp
+		}
 		j, err := json.Marshal(event)
 		if err != nil {
 			log.Warnf("failed to marshal event: %s", err.Error())
 		}
 		me.eventChannel <- string(j)
 	}
+	// Add 1 second to the last time stamp if we found records from this pull.
+	// This will prevent duplicate records from being retrieved
+	if len(events) > 0 {
+		me.saveLastRun(lastTime.Add(time.Second * 1).Format(time.RFC3339))
+	}
+
 	return nil
+
+}
+func (me *MuleEventEmitter) GetLastRun() (string, string) {
+	tStamp, _ := me.cache.Get(CacheKeyTimeStamp)
+	now := time.Now()
+	tNow := now.Format(time.RFC3339)
+	if tStamp == nil {
+		tStamp = tNow
+		me.saveLastRun(tNow)
+	}
+	return tStamp.(string), tNow
+}
+func (me *MuleEventEmitter) saveLastRun(lastTime string)  {
+	me.cache.Set(CacheKeyTimeStamp, lastTime)
+	me.cache.Save(me.cachePath)
 }
 
 // OnConfigChange passes the new config to the client to handle config changes
-// since the MuleEventEmitter does not have any config value references.
+// since the MuleEventEmitter only has cache config value references and should not be changed
 func (me *MuleEventEmitter) OnConfigChange(gatewayCfg *config.AgentConfig) {
 	me.client.OnConfigChange(gatewayCfg.MulesoftConfig)
```

Comment:
gatewayCfg.MulesoftConfig contains the cachePath variable that this struct is using, so we need to set me.cachePath here in case it changed on the config.

Reply:
Not sure about that, as it is a running job reliant upon a valid cache- I omitted this intentionally

Reply:
Ok